stdlib/list.mli

```diff
@@ -41,7 +41,12 @@ type 'a t = 'a list = [] | (::) of 'a * 'a list (**)
 (** An alias for the type of lists. *)
 
 val length : 'a list -> int
-(** Return the length (number of elements) of the given list. *)
+(** Return the length (number of elements) of the given list.
+   Note: the function is not constant time, and goes through all elements in the list,
+   in order to calculate the length.
+   To compare with another length, [compare_lengths] and [compare_length_with] are more efficient.
+   Alternatively, you can use the [buffer] type to get a constant-time [length] function.
```

Comment:
I find that a bit long winded and I'm not sure I understand what this `buffer` does here. Here is a another proposition:

```
[length l] is the number of elements in [l] (traverses the list). See 
also {!compare_length} and {!compare_length_with}.
```

Reply:
I think this doesn't warn the user enough that this is highly inefficient btw

Reply:
The `length` function is as efficient as it can be given the definition of the `list` type.

It is probably useful to remind readers (that missed the implication of this definition) that computing the length of a linked list requires to traverse the list. However, I agree that this should a short reminder.

A buffer is not a linked list, it doesn't make sense to advertise it in the `List` module. In which context, did you end up relying on `List.length`?

Reply:
#10464 (sigh) is kinda related here, too (the motivation for `List.is_empty` arising from beginner use of `List.length l = 0`.

Reply:
Maybe promoting the warning to a subordinate clause rather than a comment in parentheses work better
```
[length l] is the number of elements in [l], which is computed by traversing the list.
See also {!compare_length} and {!compare_length_with}.
```
?

Reply:
In `List.sort` we use the phrase `It runs in constant stack space`. In `Hashtbl.length` we use the phrase `It takes constant time`. Why not do similarly and write
```
[length l] is the number of elements of [l]. It runs in linear time. See also ...
```
or 
```
[length l] is the number of elements of [l]. It takes linear time. See also ...
```